src/ir_passes/new_mono.rs

```diff
@@ -0,0 +1,868 @@
+                let new_idx = self.base.add(new_param);
+                self.param_map.insert(param, new_idx);
+                new_idx
+            }
+            ir::ParamOwner::Sig => {
+                unreachable!("If a param is sig-owned, it should be resolved in the binding!")
+            }
+        }
+    }
+
+    /// Takes a self.underlying-owned param that is known to be bundle-owned and a port index owned by self.base,
+    /// creates a new param that points to the port index, and adds the param to self.base. Returns the
+    /// corresponding index
+    fn bundle_param(&mut self, param: ir::ParamIdx) -> ir::ParamIdx {
+        let ir::Param { info, .. } = self.underlying.get(param);
+        let mono_info = self.info(info);
+        let mono_owner = ir::ParamOwner::Bundle(self.curr_port.unwrap());
+
+        if let Some(new_param_idx) = self.param_map.get(&param) {
+            let mut new_param = self.base.get_mut(*new_param_idx);
+
+            new_param.owner = mono_owner;
+            new_param.info = mono_info;
+            new_param.default = None;
+
+            return *new_param_idx;
+        };
+
+        let mono_param = ir::Param {
+            owner: mono_owner,
+            info: self.info(info),
+            default: None,
+        };
+
+        let new_idx = self.base.add(mono_param);
+        self.param_map.insert(param, new_idx);
+        new_idx
+    }
+
+    /// Translates an ExprIdx defined by `underlying` to correponding one in `base`.
+    fn expr(&mut self, expr: ir::ExprIdx) -> ir::ExprIdx {
+        if let Some(idx) = self.expr_map.get(&expr) {
+            return *idx;
+        };
+        match self.underlying.get(expr).clone() {
+            ir::Expr::Param(p) => {
+                let new_idx = self
+                    .binding
+                    .get(&p)
+                    .map(|n| self.base.num(*n))
+                    .unwrap_or_else(|| self.param(p).expr(&mut self.base));
+                self.expr_map.insert(expr, new_idx);
+                new_idx
+            }
+            ir::Expr::Concrete(n) => {
+                let new_idx = self.base.num(n);
+                self.expr_map.insert(expr, new_idx);
+                new_idx
+            }
+            ir::Expr::Bin { op, lhs, rhs } => {
+                let lhs = self.expr(lhs);
+                let rhs = self.expr(rhs);
+                let new_expr = self.base.bin(ir::Expr::Bin { op, lhs, rhs });
+                let new_idx = self.base.add(new_expr);
+                self.expr_map.insert(expr, new_idx);
+                new_idx
+            }
+            ir::Expr::Fn { op, args } => {
+                let args = args.iter().map(|idx| self.expr(*idx)).collect_vec();
+                let new_expr = self.base.func(ir::Expr::Fn {
+                    op,
+                    args: Box::new(args),
+                });
+                let new_idx = self.base.add(new_expr);
+                self.expr_map.insert(expr, new_idx);
```

Comment:
Seems like all variants do the `expr_map.insert`. Let's move it out of the match?